### PR TITLE
Fixes #31532 - Regenerate entitlement certs on custom repo update for RHEL 6.8 and older

### PR DESCRIPTION
**Pull request by jeremylenz**
The "Restrict to OS Version" feature doesn't work for clients on RHEL 6.8 and older.

This fixes that by doing the following:

* When a custom repo is updated, see if the 'OS Versions' field was changed
* If so, and `rhel-6` is involved, get a list of Candlepin consumers for every client that uses RHEL 6.8 and older
* Send a request to Candlepin to regenerate the entitlement certificates for each consumer in the list

After this change, `subscription-manager refresh` should cause the certificates to be properly refreshed on RHEL 6.8 clients, including the 'Required Tags' field.  This will allow the "Restrict to OS Version" feature to work properly on these clients.

Opening this as a draft because this approach is not ideal.  Ideally, we would only send the request for the _relevant_ consumers, that is, those that are on the older OS _and_ ~~subscribe to the product containing the repo that was just changed~~.  (_update_: this would not work because of SCA) Even more ideally, we could send a single bulk request to Candlepin, but I don't think that's possible.

### CODE REVIEW

File: app/lib/actions/katello/repository/update.rb [hunk 1]
```diff
@@ -42,6 +43,24 @@ def plan(root, repo_params)
...
+                Rails.logger.info("Looking for RHEL 6.8 and older clients")
+                cp_consumers = ::Katello::Host::ContentFacet
+                  .joins(host: :operatingsystem)
+                  .where("operatingsystems.major = '6' AND operatingsystems.minor < '9' AND operatingsystems.name = 'RedHat'")
+                  .map(&:uuid)
+                # TODO: make this an action
+                Rails.logger.info("Marking entitlement certs dirty for #{pluralize(cp_consumers.length, 'client')}")
+                cp_consumers.each do |uuid|
+                  ::Katello::Resources::Candlepin::Entitlement.regenerate_entitlement_certificates_for_consumer(uuid, true)
+                end
```
Comment (jeremylenz) [inline app/lib/actions/katello/repository/update.rb:60]: If we decide to go with this approach I'll make this a Dynflow action.

File: app/lib/actions/katello/repository/update.rb [hunk 2]
```diff
@@ -42,6 +43,24 @@ def plan(root, repo_params)
...
                                        gpg_url: repository.yum_gpg_key_url)
+
+            if root.previous_changes.key?('os_versions')
+              # If Restrict to OS Version is changing _from_ or _to_ 'rhel-6'...
+              if root.previous_changes['os_versions'].any? { |ch| ch.include?('rhel-6') }
+                # ...force regeneration of entitlement certs on RHEL 6.8 and older
+                Rails.logger.info("Looking for RHEL 6.8 and older clients")
+                cp_consumers = ::Katello::Host::ContentFacet
+                  .joins(host: :operatingsystem)
+                  .where("operatingsystems.major = '6' AND operatingsystems.minor < '9' AND operatingsystems.name = 'RedHat'")
```
Comment (jeremylenz) [inline app/lib/actions/katello/repository/update.rb:54]: if there's ever a RHEL 6.10 this `<` will break

Comment (jlsherrill) [reply app/lib/actions/katello/repository/update.rb:54]: Another option would be to look for hosts that have a subscription manager installed that is older than what works.  That would be more accurate IMO

File: app/lib/actions/katello/repository/update.rb (hunk 1 above)
Comment (jlsherrill) [reply app/lib/actions/katello/repository/update.rb:60]: +1 and it should spawn off another task, not run within this one.  We do this for example here:  https://github.com/Katello/katello/blob/master/app/lib/actions/katello/repository/update.rb#L62

File: app/lib/actions/katello/repository/update.rb (hunk 2 above)
Comment (jeremylenz) [reply app/lib/actions/katello/repository/update.rb:54]: Yeah that would be ideal but how would I know the installed sub-man version for all my hosts?

**Comment by jeremylenz** [comment]
Updated:
* add new actions `Actions::Katello::Consumer::RegenerateEntitlementCertificate(s)`
* move the code that actually calls Candlepin to new methods in `Katello::Candlepin::Consumer` and `Katello::Resources::Candlepin::Consumer`
* update `Actions::Katello::Repository::Update` to spawn the new action using `async_task`
* clean up the above action to avoid calling `plan_self` inside a conditional

**Comment by jeremylenz** [comment]
Updated to run the action in batches of 10 hosts each (speeds up performance by 10x)

**Comment by jeremylenz** [comment]
Closing after discussion with Rich Jerrido and @mccun934.  Support for older RHEL 6 clients is not a requirement; these users should upgrade at least to RHEL6.latest.

**Pull request closed without merging**